### DISCUSSION TITLE
Buffer de tamaño "infinito"

### DISCUSSION THREAD

**Discussion by julchat** [dds-jv/foro#135] [Consultas de objetos] [answered]
Hola, uno de los requerimientos adicionales es poder leer de una cualquier archivo de cualquier tamaño, que podría ser infinito. 
Lo que había pensado en primera instancia es tener una clase Archivo donde le delegue la responsabilidad de decirme cuantos bytes tiene el archivo, mensaje que mandaría desde la clase que utiliza el usuario, y con ese resultado declarar el array en el buffer con dicho tamaño (esto no se si en java se puede, estoy acostumbrado a que el tamaño del array se defina explícitamente previo a la ejecución del programa), pero también me acordé un poco de lo que habíamos hablado sobre el patrón del stream, donde yo podría ir metiendo los caracteres del archivo uno a uno, y cuando sea la hora de decirle al LowLevelFileSystem que lea el archivo, hacer algo así como un len del stream.

**Comment by gastonprieto** [comment]
por las dudas, ya lo vi, mas tarde (o sea a la mañana te respondo)

**Comment by julchat** [comment]
Profe creo que se olvidó, o quizás esta ocupado, aviso por las dudas

**Comment by gastonprieto** [comment]
Buenas!

> Lo que había pensado en primera instancia es tener una clase Archivo donde le delegue la responsabilidad de decirme cuantos bytes tiene el archivo, mensaje que mandaría desde la clase que utiliza el usuario, y con ese resultado declarar el array en el buffer con dicho tamaño (esto no se si en java se puede, estoy acostumbrado a que el tamaño del array se defina explícitamente previo a la ejecución del programa), 

Si, puede declararse de forma dinamica el array, pero bueno el chiste de eso es que no tengas todo el archivo a memoria de una, lamentablemente no tenemos memoria infinita 😢 Así que tendriamos que buscar otra opción

> pero también me acordé un poco de lo que habíamos hablado sobre el patrón del stream, donde yo podría ir metiendo los caracteres del archivo uno a uno, y cuando sea la hora de decirle al LowLevelFileSystem que lea el archivo, hacer algo así como un len del stream

<details>
<summary>Creo que lo que estas encarando al reves, la idea del stream es que te genere valores (en este caso podrian ser bloques de tamaño fijo), la forma de generar esos valores en este caso va a ser leyendo del lowlevelfilesystem y pasandoselos a algo. Si desplegas el details vas a avanzar por este camino, pero tal vez es el mas complejo
</summary>

Por ejemplo una posible implementación de un stream generico sincronico (ojo, esto potencialemente también podría ser asincronico) podría ser:

```
class StreamGenerico
    public void each(operacion) { 
    	while(this.tengoMasValoresAEmitir()) { operacion(this.generarValorNuevo()) } 
    }
```

y lo usarías como

streamGenerico.each( nro => System.out.println( "recibi el nro " + nro) )

Tal vez esto te da una idea de como podrías implementar algo similar al stream de Java

Para ver como hacerlo posta en Java, tal vez te sirve: https://stackoverflow.com/questions/30685623/how-to-implement-a-java-stream

</details>

----

## Pensando otras alternativas

* Por otro lado no hace falta que lo encares de esta forma, hay otra cosa que tal vez podría servirte, que pasaría si modelas algo que sepas donde estas parado en el archivo? y puedas ir avanzando las lecturas de a poco? se te ocurre cual sería la interfaz?

* Qué pasaría si vos le decís que hacer cada vez que lee un bloque y haces algo que recorra el archivo por vos, como si fuera un #map de una lista, pero sería algo que sabe en realidad recorrer un archivo con tamaños fijos.

**Comment by julchat** [comment]
`Por otro lado no hace falta que lo encares de esta forma, hay otra cosa que tal vez podría servirte, que pasaría si modelas algo que sepas donde estas parado en el archivo? y puedas ir avanzando las lecturas de a poco? se te ocurre cual sería la interfaz?`
- Uh, yo pensaba que el que se encargaba de pararse en una parte del archivo y moverse a medida que haces las lecturas era el fd o flujo, como se manejaba, ni idea, pensaba que era implementación del LowLevelFileSystem. Me habré confundido porque estoy acostumbrado a usar stdio.h de c, donde vos pasas el puntero FILE que ya te apunta a "donde habias dejado" al archivo.

Es más, si me decís eso, si yo quisiera implementarlo desde la parte que diseño, cuando le mande la operación al LowLevelFileSystem y quiero empezar a leer desde X bytes empezados del archivo, no tengo forma de hacerlo, fuera de algo rebuscado como escribir un archivo auxiliar que comience desde lo que quería comenzar a leer

`Qué pasaría si vos le decís que hacer cada vez que lee un bloque y haces algo que recorra el archivo por vos, como si fuera un #map de una lista, pero sería algo que sabe en realidad recorrer un archivo con tamaños fijos.`

Esto no lo entendí muy bien, osea entiendo la idea de alguna forma poder recorrer un archivo infinito, saber donde está lo que quiero, ir recorriendo hasta que llego a lo que quiero, pero no entiendo que sería esta lista, lista de caracteres o bytes? La ventaja de trabajar a los bytes como una lista en vez de un array cual sería? poder hacerlos stream y operar con ellos? Pero aún así caigo en el problema del item anterior de que no tengo forma de decirle al LowLevelFileSystem, "leete desde acá", o "lee esto primero, y desde ahi lee esto"

Si querés podemos charlarlo mañana en clase también, aunque me queda mas grabado si lo puedo leer.

**Comment by gastonprieto** [comment]
> Uh, yo pensaba que el que se encargaba de pararse en una parte del archivo y moverse a medida que haces las lecturas era el fd o flujo, como se manejaba, ni idea, pensaba que era implementación del LowLevelFileSystem. Me habré confundido porque estoy acostumbrado a usar stdio.h de c, donde vos pasas el puntero FILE que ya te apunta a "donde habias dejado" al archivo.

Ah.. si se puede interpretar así y tiene sentido por los nombres que maneja. 
Yo cuando te respondi en realidad pense en las [implementaciones de filesystem](https://libfuse.github.io/doxygen/structfuse__operations.html#a2a1c6b4ce1845de56863f8b7939501b5) que suelen recibir inicio y fin de lo que queres leer. 

Pero tu interpretación tiene mucho sentido. Y por cucaracha me estan diciendo que era la idea original jaja Pero que se puede ir por ambos caminos.

Creo que si vas por mi camino te respondo un par de las siguientes preguntas. que hiciste. Pero no entiendo la comparativa que haces entre lista y array

**Comment by gastonprieto** [comment]
Igual si vas por tu lado, ese coso que sabe hasta donde leiste es el lowlevelfilesystem, asi que ya tenes una parte resulta, te queda ver como aprovechar ese estado que tiene el lowlevelfilesystem

**Comment by gastonprieto** [comment]
@julchat  sirvio esto ? te quedo alguna duda mas ?

**Comment by julchat** [comment, marked as answer]
Sisi me sirvio mucho,
muchas gracias
